src/scry/protocol/response_error.cr

```diff
@@ -15,15 +15,33 @@ module Scry::Protocol
     RequestCancelled = -32800
   end
 
+  class ProtocolException < Exception
+    getter code
+
+    def initialize(
+      @message : String? = nil,
+      @cause : Exception? = nil,
+      @code : ErrorCodes = ErrorCodes::UnknownErrorCode
+    )
+    end
+  end
+
   struct ResponseError
     JSON.mapping({
-      code:    Int32,
+      code:    ErrorCodes,
       message: String,
       data:    Array(String)?,
     }, true)
 
-    def initialize(@message, @data)
-      @code = ErrorCodes::UnknownErrorCode.value
+    def initialize(@message, @data = nil, @code = ErrorCodes::UnknownErrorCode)
+    end
+
+    def initialize(ex : Exception)
```

Comment:
I suggest to keep only one `initialize` constructor, and have multiple overload of `.new` with an `Exception` or `ProtocolException` param

Reply:
When would I want to override one versus the other?  I come from java land so I'm not too familiar with this new vs initialize concept

Reply:
basically it is preferred to have one `initialize` which will initialize all instance variables (for this `initialize`, the compiler will generate a class method `new` with the same signature), then you overload the `new` for other kind of object creation (from exception instead of message/data/code)

So you'd keep the top `initialize` method, then do:
```cr
def self.new(ex : Exception)
  # delegate to the `new`, generated for the above `initialize`
  new(ex.message || "Unknown error", ex.backtrace?)
end
```

same logic for the other one